SSA/Projects/InstCombine/scripts/mlir-tool.py

```diff
@@ -0,0 +1,183 @@
+        return f.read()
+
+
+for file in os.listdir(directory):
+    filename = os.fsdecode(file)
+    print(filename)
+    stem = "g" + filename.split(".")[0].replace("-", "h")
+    output = ""
+    data1 = read_file(os.path.join("../vcombined-mlir"))
+    data2 = read_file(os.path.join("../vbefore-mlir"))
+    parser1 = Parser(ctx, data1)
+    try:
+        module1 = parser1.parse_module()
+        parser2 = Parser(ctx, data2)
+        module2 = parser2.parse_module()
+        funcs = [
+            func
+            for func in module1.walk()
+            if isinstance(func, FuncOp)
+            and all(allowed(o) for o in func.walk())
+            and size(func) > 1
+        ]
+        for func in funcs:
+            try:
+                other = next(
+                    f
+                    for f in module2.walk()
+                    if isinstance(f, FuncOp) and f.sym_name == func.sym_name
+                )
+                if all(allowed(o) for o in other.walk()):
+                    s1 = showr(func.body)
+                    s2 = showr(other.body)
+                    name = func.sym_name.data
+                    if s1 == s2:
+                        continue
+                    if "vector" in (s1 + s2):
+                        continue
+                    print(f"-----{filename}.{func.sym_name}-----")
+                    o1 = f"""
+def {name}_before := [llvm|
+{s2}
+]
+def {name}_after := [llvm|
+{s1}
+]
+theorem {name}_proof : {name}_before ⊑ {name}_after := by
+unfold {name}_before {name}_after
+simp_alive_peephole
+simp_alive_undef
+simp_alive_ops
+simp_alive_case_bash
+try alive_auto
+---BEGIN {name}
+all_goals (try extract_goal ; sorry)
+---END {name}\n\n\n"""
+                    print(o1)
+                    write_file = os.path.join(
+                        "../lean-mlir",
+                        "SSA",
+                        "Projects",
+                        "InstCombine",
+                        "tests",
+                        "LLVM",
+                        f"{stem}.lean",
+                    )
+                    with open(write_file, "a+") as f3:
+                        if os.stat(write_file).st_size == 0:
+                            f3.write(
+                                """
+import SSA.Projects.InstCombine.LLVM.PrettyEDSL
+import SSA.Projects.InstCombine.TacticAuto
+import SSA.Projects.InstCombine.LLVM.Semantics
+open LLVM
+open BitVec
+
+open MLIR AST
+open Std (BitVec)
+open Ctxt (Var)
+
+set_option linter.deprecated false
+set_option linter.unreachableTactic false
+set_option linter.unusedTactic false
+                                                                """
+                            )
+                        f3.write(o1)
+
+            except StopIteration as e:
+                print(
+                    f"Cannot find a function named {func.sym_name} in file {filename}"
+                )
+    except ParseError as e:
+        print("failed to parse the file")
```

Comment:
It's good practice to make the try/catch code close to each other, here the only place it could throw is in the first few lines, might as well dedent after that and do the loop below the except